corehq/apps/app_manager/models.py

```diff
@@ -3864,162 +3864,6 @@ def super_save():
             super_save()
 
 
-class VersionedDoc(LazyBlobDoc):
```

Comment:
So I'm not totally in love with this change, since `ApplicationBase` is already 600 lines long, and a VersionedDoc is a relatively clear concept (although the ODK urls don't belong in it...). But I was finding it hard to work with (because VersionedDoc controls copy_of but ApplicationBase controls all the other build-related fields). I'd rather collapse the unused level of hierarchy and instead break ApplicationBase up using composition.

Open to feedback. @proteusvacuum @emord you've also thought about the app manager models hierarchy.

Reply:
I would push for keeping `VersionedDoc`, retaining version related logic in it.

And for build related things to go into application or if possible, in a different class and then that is used to make build related things instead of populating `Application` class. Inheriting `Application` from it would be the easier way out but let's avoid that. I have extracted classes like `HostedCCZUtility` which dealt with creating the CCZ file for `HostedCCZ` and related operations (the build in this case for Application) and then use that to handle operations.

Reply:
Looked at this a bit more, and I still like the idea of making VersionedDoc a separate utility class, but doing that is more complicated than I expected. VersionedDoc isn't a clean abstraction, given its use of application-specific urls (short_odk_url, etc) and the `app_manager/applications` couch view.

I don't think it's worth the effort to clean up VersionedDoc enough to make it truly generic and pull it out of the Application hierarchy, and I still think the extra layer of hierarchy (e.g., having two implementations of `make_build`) adds more in complexity than it's worth. I'm inclined to leave this code as it is in this PR, removing VersionedDoc altogether.